File: LinkedList/ReverseLinkedList/ReverseLinkedList.cpp
/// 迭代
class Solution {
public:
    ListNode* reverseList(ListNode *head) {
        ListNode *prev = nullptr;
        ListNode *back;
        while(head) {
            back = head->next;
            head->next = prev;
            prev = head;
            head = back;
        }
        return prev;
    }
};


/// 递归一：
class Solution {
public:
    ListNode* _reverseList(ListNode *head, ListNode *prev = nullptr){
        if (head == nullptr) { return prev; }
        ListNode *back = head->next;
        head->next = prev;
        return _reverseList(back, head);
    }
    
    ListNode* reverseList(ListNode* head) {
        return _reverseList(head);
    }
};


/// 递归二：
class Solution {
public:
    ListNode* reverseList(ListNode *head) {
        if(head == nullptr || head->next == nullptr) { return head; }
        ListNode *new_head = reverseList(head->next);
        head->next->next = head;
        head->next = nullptr;
        return new_head;
    }
};

File: LinkedList/ReverseLinkedList/ReverseLinkedList.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution
{


    /**
     * @param ListNode $head
     * @return ListNode
     */
    function reverseList($head)
    {
        $current = $head;
        $newList = null;
        while ($current) {
            $nextNode = $current->next;
            $current->next = $newList;
            $newList = $current;
            $current = $nextNode;
        }

        return $newList;
    }


    //recursion
    function reverseListForRecursion($head){

        if(!$head || !$head->next){
            return $head;
        }
        $newList=$this->reverseList($head->next);
        //we want n(k+1)'s next node to point k(n),so
        $head->next->next=$head;
        $head->next=null;
        return $newList;
    }
}